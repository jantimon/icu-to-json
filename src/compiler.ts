import { parse, TYPE } from "@formatjs/icu-messageformat-parser";
import type {
  MessageFormatElement,
  NumberSkeleton,
} from "@formatjs/icu-messageformat-parser";
import {
  CompiledArgument,
  CompiledAst,
  CompiledAstContents,
  CompiledFn,
  CompiledOrdinal,
  CompiledPlural,
  CompiledTag,
  JSON_AST_TYPE_FN,
  JSON_AST_TYPE_PLURAL,
  JSON_AST_TYPE_SELECT,
  JSON_AST_TYPE_SELECTORDINAL,
  JSON_AST_TYPE_TAG,
} from "./constants.js";

type Ast = ReturnType<typeof parse>;

interface CompileOptions {
  /** Allow string interpolations like "Hello [0]!" format */
  allowStringInterpolation?: boolean;
}

export const compileToJson = (
  str: string,
  options?: CompileOptions
): CompiledAst => compile(str, options).json;

/**
 * Convert an ICU messageformat string to a json ast which can be render by the icu-to-json runtime.
 */
export const compile = (str: string, options?: CompileOptions) => {
  const ast = parse(str);
  const args = getAllArguments(ast);
  const result = compileAst(ast) satisfies CompiledAst;
  // to reduce the size of the json payload,
  // pure text is returned as a string to save the surrounding array brackets
  const json: CompiledAst =
    result.length === 1 && typeof result[0] === "string" ? result[0] : result;
  // String interpolation is not supported by format.js
  if (typeof json === "string" && options?.allowStringInterpolation) {
    return compileStringInterpolation(json);
  }
  return {
    args,
    json,
  };
};

const compileAst = (ast: Ast): CompiledAstContents[] => {
  return ast.map((node): CompiledAstContents => {
    switch (node.type) {
      case TYPE.literal:
        return node.value;
      case TYPE.argument:
        return [node.value];
      case TYPE.plural:
      case TYPE.select:
        return [
          node.value,
          node.type === TYPE.select
            ? JSON_AST_TYPE_SELECT
            : // in format.js
            // there is no selectordinal type
            // selectordinal is a plural with pluralType = "ordinal"
            node.pluralType === "ordinal"
            ? JSON_AST_TYPE_SELECTORDINAL
            : JSON_AST_TYPE_PLURAL,
          Object.fromEntries(
            Object.entries(node.options).map(([caseName, pluralCase]) => {
              // add support for
              // {{count, plural, one {#} =12{a bunch} other {#s}}}
              return [
                caseName.replace(/^\s*=/, ""),
                compileAst(pluralCase.value),
              ];
            })
          ),
        ] satisfies CompiledPlural;
      case TYPE.tag:
        return [
          node.value,
          JSON_AST_TYPE_TAG,
          ...compileAst(node.children),
        ] satisfies CompiledTag;
      case TYPE.pound:
        // TYPE.pound is the octothorpe character used in plural and selectordinal
        return CompiledOrdinal;
      case TYPE.time:
        return [
          node.value,
          JSON_AST_TYPE_FN,
          "time",
          ...(node.style !== null ? [node.style] : []),
        ] satisfies CompiledFn;
      case TYPE.date:
        return [
          node.value,
          JSON_AST_TYPE_FN,
          "date",
          ...(node.style !== null ? [node.style] : []),
        ] satisfies CompiledFn;
      case TYPE.number:
        return node.style === null
          ? ([
              node.value,
              JSON_AST_TYPE_FN,
              "number",
            ] satisfies CompiledFn)
          : ([
             node.value,
             JSON_AST_TYPE_FN,
              "numberFmt",
              getNumberFormat(node.style),
            ] satisfies CompiledFn);
      default:
        console.log(node);
        throw new Error("Not implemented");
    }
  });
};

export type ArgumentUsage =
  | "argument"
  | "tag"
  | "select"
  | "selectordinal"
  | "number"
  | "numberFmt"
  | "date"
  | "time"
  | "plural";

const getAllArguments = (ast: Ast) => {
  const args: Record<string, ArgumentUsage[]> = {};
  const getArgs = (node: MessageFormatElement) => {
    switch (node.type) {
      case TYPE.literal:
      case TYPE.pound:
        break;
      case TYPE.number:
        if (node.style === null) {
          args[node.value] = addIfNotExists(args[node.value], "number");
        } else {
          args[node.value] = addIfNotExists(args[node.value], "numberFmt");
        }
        break;
      case TYPE.date:
        args[node.value] = addIfNotExists(args[node.value], "date");
        break;
      case TYPE.time:
        args[node.value] = addIfNotExists(args[node.value], "time");
        break;
      case TYPE.tag:
        args[node.value] = addIfNotExists(args[node.value], "tag");
        node.children.forEach(getArgs);
        break;
      case TYPE.plural:
        args[node.value] = addIfNotExists(
          args[node.value],
          node.pluralType === "cardinal" ? "plural" : "selectordinal"
        );
        Object.entries(node.options).forEach(([, pluralCase]) => {
          pluralCase.value.forEach(getArgs);
        });
        break;
      case TYPE.select:
        args[node.value] = addIfNotExists(args[node.value], "select");
        Object.entries(node.options).forEach(([, pluralCase]) => {
          pluralCase.value.forEach(getArgs);
        });
        break;
      case TYPE.argument:
        args[node.value] = addIfNotExists(args[node.value], "argument");
        break;
      default:
        assertNever(node);
    }
  };
  ast.forEach(getArgs);
  return args;
};

/**
 * Compile string interpolations like "Hello [0]!" format
 * to a json ast.
 */
const compileStringInterpolation = (
  str: string
): { args: Record<string, ArgumentUsage[]>; json: CompiledAst } => {
  const translationParts = str
    .split(/\[(\d+)\]/g).map((part, index) => index % 2 === 0 ? part : [parseInt(part, 10)] satisfies CompiledArgument);
  if (translationParts.length === 1) {
    return {
      args: {},
      json: str,
    };
  }
  const argList = [...new Set(translationParts.filter((_, index) => index % 2 === 1))];
  const contents = translationParts satisfies CompiledAstContents[];
  return {
    args: Object.fromEntries(argList.map((arg) => [arg, ["argument"]])),
    json: contents satisfies CompiledAst,
  };
};

function getNumberFormat(style: string | NumberSkeleton | undefined) {
  if (style === "%") {
    return "percent";
  }
  return style;
}

function addIfNotExists<T>(array: T[] | undefined, item: T) {
  if (array === undefined) {
    return [item];
  }
  if (!array.includes(item)) {
    array.push(item);
  }
  return array;
}

function assertNever(x: never) {
  throw new Error("Unexpected object: " + x);
}
