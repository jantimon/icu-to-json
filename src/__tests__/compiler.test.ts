import { compileToJson } from "../compiler.js";
import { test, expect } from "vitest";
import messages from "./__fixtures__/messages.json";

test.each(Object.entries(messages.en))(`measure size for %s example`, (testName, icuMessage) => {
    const json = compileToJson(icuMessage);
    const size = JSON.stringify(icuMessage).length;
    const jsonSize = JSON.stringify(json).length;
    const percent = Math.round((jsonSize / size) * 100) - 100;

    expect(
        [icuMessage, json, `\n` +
        `   size (icu message): ${size}b\n` +
        `   size (json):        ${jsonSize}b\n` +
        `   overhead:           ${percent > 0 ? "+":""}${percent}%\n`
    ]).toMatchSnapshot()
  });
