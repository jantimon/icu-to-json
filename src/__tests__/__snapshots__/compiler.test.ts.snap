// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`measure size for fn example 1`] = `
[
  "It is now {currentTime, time} on {currentTime, date}",
  [
    "It is now ",
    [
      "currentTime",
      4,
      "time",
    ],
    " on ",
    [
      "currentTime",
      4,
      "date",
    ],
  ],
  "
   size (icu message): 54b
   size (json):        71b
   overhead:           +31%
",
]
`;

exports[`measure size for integer example 1`] = `
[
  "The integer value of {amount} is {amount, number, integer}",
  [
    "The integer value of ",
    [
      "amount",
    ],
    " is ",
    [
      "amount",
      4,
      "numberFmt",
      "integer",
    ],
  ],
  "
   size (icu message): 60b
   size (json):        78b
   overhead:           +30%
",
]
`;

exports[`measure size for interpolatedStrings example 1`] = `
[
  "Hello, my name is [0] and I am [1] years old.",
  [
    "Hello, my name is ",
    [
      0,
    ],
    " and I am ",
    [
      1,
    ],
    " years old.",
  ],
  "
   size (icu message): 47b
   size (json):        57b
   overhead:           +21%
",
]
`;

exports[`measure size for money example 1`] = `
[
  "Buy now to save {amount, number, currency:GBP}",
  [
    "Buy now to save ",
    [
      "amount",
      4,
      "numberFmt",
      "currency:GBP",
    ],
  ],
  "
   size (icu message): 48b
   size (json):        60b
   overhead:           +25%
",
]
`;

exports[`measure size for number example 1`] = `
[
  "I have {numCats, number} cats.",
  [
    "I have ",
    [
      "numCats",
      4,
      "number",
    ],
    " cats.",
  ],
  "
   size (icu message): 32b
   size (json):        43b
   overhead:           +34%
",
]
`;

exports[`measure size for percentage example 1`] = `
[
  "You reached {value, number, %} of your goal!",
  [
    "You reached ",
    [
      "value",
      4,
      "numberFmt",
      "percent",
    ],
    " of your goal!",
  ],
  "
   size (icu message): 46b
   size (json):        67b
   overhead:           +46%
",
]
`;

exports[`measure size for plural example 1`] = `
[
  "{count, plural, one {# image} other {# images} }",
  [
    [
      "count",
      2,
      {
        "one": [
          0,
          " image",
        ],
        "other": [
          0,
          " images",
        ],
      },
    ],
  ],
  "
   size (icu message): 50b
   size (json):        56b
   overhead:           +12%
",
]
`;

exports[`measure size for select example 1`] = `
[
  "Hello, Your friend {friend} is now online. {gender, select, female {She} male {He} other {They}} added a new image to the system.",
  [
    "Hello, Your friend ",
    [
      "friend",
    ],
    " is now online. ",
    [
      "gender",
      1,
      {
        "female": [
          "She",
        ],
        "male": [
          "He",
        ],
        "other": [
          "They",
        ],
      },
    ],
    " added a new image to the system.",
  ],
  "
   size (icu message): 131b
   size (json):        152b
   overhead:           +16%
",
]
`;

exports[`measure size for selectNumeric example 1`] = `
[
  "She has {children, plural, =0 {no children} =1 {1 child} other {# children}} and lives in {city}.",
  [
    "She has ",
    [
      "children",
      2,
      {
        "0": [
          "no children",
        ],
        "1": [
          "1 child",
        ],
        "other": [
          0,
          " children",
        ],
      },
    ],
    " and lives in ",
    [
      "city",
    ],
    ".",
  ],
  "
   size (icu message): 99b
   size (json):        119b
   overhead:           +20%
",
]
`;

exports[`measure size for selectordinal example 1`] = `
[
  "You finished {place, selectordinal, one {#st} two {#nd} few {#rd} other {#th}}!",
  [
    "You finished ",
    [
      "place",
      3,
      {
        "few": [
          0,
          "rd",
        ],
        "one": [
          0,
          "st",
        ],
        "other": [
          0,
          "th",
        ],
        "two": [
          0,
          "nd",
        ],
      },
    ],
    "!",
  ],
  "
   size (icu message): 81b
   size (json):        97b
   overhead:           +20%
",
]
`;

exports[`measure size for tags example 1`] = `
[
  "Wow formatJs allows <b>{dynamic} tags</b>",
  [
    "Wow formatJs allows ",
    [
      "b",
      5,
      [
        "dynamic",
      ],
      " tags",
    ],
  ],
  "
   size (icu message): 43b
   size (json):        52b
   overhead:           +21%
",
]
`;

exports[`measure size for text example 1`] = `
[
  "Hello",
  "Hello",
  "
   size (icu message): 7b
   size (json):        7b
   overhead:           0%
",
]
`;

exports[`measure size for time example 1`] = `
[
  "Sale begins {start, time, short} at {start, date, short}",
  [
    "Sale begins ",
    [
      "start",
      4,
      "time",
      "short",
    ],
    " at ",
    [
      "start",
      4,
      "date",
      "short",
    ],
  ],
  "
   size (icu message): 58b
   size (json):        77b
   overhead:           +33%
",
]
`;

exports[`measure size for variable example 1`] = `
[
  "{name} has just entered the chat",
  [
    [
      "name",
    ],
    " has just entered the chat",
  ],
  "
   size (icu message): 34b
   size (json):        39b
   overhead:           +15%
",
]
`;
